Gains dataset `abcd_demo`, begin deprecation of `ald_demo`

jdhoffa opened the pull request:
In this PR I begin the depracation process of the dataset `ald_demo`, in favour of the dataset `abcd_demo`. This has been done by: 
- Adding a new dataset, `abcd_demo`, that contains identical data to `ald_demo`
- Adding a `superseded` badge to the documentation of `ald_demo`

I wanted to also add a warning to `ald_demo` (using `lifecycle::deprecate_soft()`), however I wasn't sure how to do this, since we are working with a dataset, and not a function (there doesn't seem to be anywhere to embed the call to `lifecycle::deprecate_soft()`). I also couldn't find any guidance on the lifecycle website, @maurolepore any comments? 

Closes #279

maurolepore commented:
@jdhoffa I added a warning via `delayAssign()`. Note I removed `ald_demo` from data/ because we no longer need it: We assign `ald_demo <- abcd_demo` on the fly via `delayAssign()`.

Feel free to remove my commits. I now realized I should have created a PR off your PR. Let me know if you prefer me to revert and do that instead.

``` r
library(r2dii.data)

# Note the warning
identical(ald_demo, abcd_demo)
#> Warning: `ald_demo` was superseded in r2dii.data 0.3.0. Please use `abcd_demo`
#> instead.
#> [1] TRUE

# After the first warning in a session, it doesn't warn 
# I don't know why but it's kind of nice
identical(ald_demo, abcd_demo)
#> [1] TRUE
```

<sup>Created on 2022-04-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

jdhoffa commented:
> @jdhoffa I added a warning via `delayAssign()`. Note I removed `ald_demo` from data/ because we no longer need it: We assign `ald_demo <- abcd_demo` on the fly via `delayAssign()`.
> 
> Note the warning assumes the next release will have version 0.3.0. Feel free to adapt.
> 
> Feel free to remove my commits. I now realized I should have created a PR off your PR. Let me know if you prefer me to revert and do that instead.
> 
> ```r
> library(r2dii.data)
> 
> # Note the warning
> identical(ald_demo, abcd_demo)
> #> Warning: `ald_demo` was superseded in r2dii.data 0.3.0. Please use `abcd_demo`
> #> instead.
> #> [1] TRUE
> 
> # After the first warning in a session, it doesn't warn 
> # I don't know why but it's kind of nice
> identical(ald_demo, abcd_demo)
> #> [1] TRUE
> ```
> 
> Created on 2022-04-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)

Thanks @maurolepore this works awesome. And that's ok that you pushed directly to this PR. 
I do plan on this change being bundled in `r2dii.data 0.3.0` so that works well. 

I think this PR is ready for a proper review now then.

maurolepore commented on DESCRIPTION:
```diff
@@ -38,6 +28,7 @@ BugReports: https://github.com/2DegreesInvesting/r2dii.data/issues
 Depends: 
     R (>= 3.4)
 Imports:
+    lifecycle,
```
Is it still worth depending on lifecycle? The warning we implemented in `delayAssign()` does not use it.

jdhoffa replied:
I think we still need to depend on it for the badges it provides??

maurolepore replied:
Yeah, the badges are nice. Maybe once you complete the deprecation you may even remove the dependency on lifecycle. (So far this package imports nothing that does not come with R. There are a few suggested packages but nothing in imports other than utils and stats.)

jdhoffa replied:
Ya, that's a great idea.

maurolepore commented on R/ald_demo.R:
```diff
@@ -1,16 +1,19 @@
 #' An asset level dataset for demonstration
 #'
 #' @description
+#' `r lifecycle::badge("superseded")`
```
Yeah, this IS nice but if it's the only place we use lifecycle maybe it's not worth gaining a dependency.

jdhoffa replied:
So, knowing that we will be introducing several other `lifecycle` dependencies in `r2dii.match` and `r2dii.analysis` (mainly for the badges on arguments that will be lost), I am not terribly opposed to adding it here as well. 

`r2dii.data` is basically useless without those packages as well, so they will need to install that dependency at some point.

maurolepore replied:
Good point.

maurolepore commented on R/ald_demo.R:
```diff
@@ -1,16 +1,19 @@
 #'
 #' @template info_demo-datasets
 #'
+#' @keywords internal
```
Maybe we can delete R/ald_demo.R and use `#' @aliases` in `abcd_demo`?

jdhoffa replied:
Hmm, ok so two options: 
* Keep `R/ald_demo.R` and `R/abcd_demo.R`. Benefit here is that, we can keep the documentation of `ald_demo` internal (so it is hard to find), but if someone does try to pull up the documentation directly, they will see immediately that the dataset is superseded (by the lifecycle badge) 
* Make `ald_demo` an alias of `R/abcd_demo.R`. Benefit here is that there is less code relating to `ald_demo`. Downside is that if people run `?ald_demo`, they wont be notified that it is superseded, and they may not even notice that they are looking at the documentation to `abcd_demo`.

jdhoffa replied:
Personally I tend to prefer that if people lookup `?ald_demo` they get a big fat warning that it is superseded... but I'm not opposed to the `@alias` approach if you think it is cleaner.

jdhoffa replied:
Ok, I've minimized the documentation, but still using the lifecycle bade. Let me know what you think.

maurolepore commented on data-raw/abcd_demo.R:
```diff
@@ -168,7 +168,9 @@ ordered_names <- c(
   "ald_timestamp",
   emission_factor_unit = "ald_emission_factor_unit"
 )
+# FIXME: Remove all references to abcd_demo from this file
```
SORRY! I meant to say 

FIXME: Remove all references to `ald_demo` from this file. 

We no longer need `ald_demo`  -- it's assigned on the fly if and when the user calls `ald_demo`.

maurolepore commented on data/ald_demo.R:
```diff
@@ -0,0 +1,14 @@
+    warning(
+      "`ald_demo` was superseded in r2dii.data 0.3.0. ",
+      "Please use `abcd_demo` instead.",
+      "\n Read this blog post to learn more:",
+      "\n https://2degreesinvesting.github.io/posts/2022-03-02-ald-becomes-abcd/",
```
Running this in my head I think these two lines will be indented by one character. Is this what you want?

jdhoffa replied:
The `\n ` character (with the space) does not indent the text anymore than the rest of the warning: 
<img width="635" alt="Screenshot 2022-04-28 at 14 45 11" src="https://user-images.githubusercontent.com/8741309/165754909-a621c41b-ca92-4e8e-837d-e643700cccae.png">

maurolepore reviewed (approved):
LGTM but see my comments. In general I would do one more pass looking for things to prune -- so we don't leave things behind. For example, we no longer need to export `ald_demo` so there is no point in creating it in data-raw/ or having a .csv file.

jdhoffa commented:
> LGTM but see my comments. In general I would do one more pass looking for things to prune -- so we don't leave things behind. For example, we no longer need to export `ald_demo` so there is no point in creating it in data-raw/ or having a .csv file.

Ok fair enough, I will. If I'm honestly, I've been a little conflicted about how to handle this update. 
Part of me wants to keep the PRs very focused and small (ie. ONLY add `abcd_demo`, and warn that `ald_demo` will be removed, and not change much else). The other side of me wants to be complete, and remove EVERYTHING related to `ald_demo` except for whatever is necessary to keep it working until it is fully deprecated. 

It sounds like you prefer the latter approach.

jdhoffa commented:
> For example, we no longer need to export `ald_demo` so there is no point in creating it in data-raw/ or having a .csv file.

Also it isn't created in `data-raw/` and there is no `.csv` anymore? These had already been removed. The only place it appears still is in the data-dictionary definition, which i will remove it from.

maurolepore commented:
Yeah, I think your idea of keeping the PR small is totally valid. Two advantages of the alternative that I see:

1. Get closer to meeting the legal obligation of removing references to "ald"
2. Get stuff done while you have the context in your head. No big deal -- as you can search easily for file-names and text containing "ald".
